tests/xdg_popup.cpp

```diff
@@ -732,19 +763,49 @@ TEST_P(XdgPopupTest, grabbed_popup_gets_done_event_when_new_toplevel_created)
     // This is needed to get a serial, which will be used later on
     pointer.move_to(manager->window_x + 2, manager->window_y + 2);
     pointer.left_click();
-    manager->client.roundtrip();
+    manager->client->roundtrip();
 
     auto positioner = PositionerParams{}
         .with_size(30, 30)
         .with_anchor(XDG_POSITIONER_ANCHOR_TOP_LEFT)
         .with_gravity(XDG_POSITIONER_GRAVITY_BOTTOM_RIGHT)
         .with_grab();
     manager->map_popup(positioner);
-    manager->client.roundtrip();
+    manager->client->roundtrip();
 
     EXPECT_CALL(*manager, popup_done());
 
-    manager->client.create_visible_surface(window_width, window_height);
+    manager->client->create_visible_surface(window_width, window_height);
+}
+
+TEST_P(XdgPopupTest, grabbed_popups_get_done_events_in_correct_order)
+{
+    auto const& param = GetParam();
+    auto top_popup_manager = param.build(this);
+    auto pointer = the_server().create_pointer();
+
+    // This is needed to get a serial, which will be used later on
+    pointer.move_to(top_popup_manager->window_x + 2, top_popup_manager->window_y + 2);
+    pointer.left_click();
+    top_popup_manager->client->roundtrip();
+
+    auto positioner = PositionerParams{}
+        .with_size(30, 30)
+        .with_anchor(XDG_POSITIONER_ANCHOR_TOP_LEFT)
+        .with_gravity(XDG_POSITIONER_GRAVITY_BOTTOM_RIGHT)
+        .with_grab();
+    top_popup_manager->map_popup(positioner);
+    top_popup_manager->client->roundtrip();
+    auto sub_popup_manager = top_popup_manager->child();
+    sub_popup_manager->map_popup(positioner);
+    top_popup_manager->client->roundtrip();
+
+    InSequence seq;
+    EXPECT_CALL(*sub_popup_manager, popup_done());
+    EXPECT_CALL(*top_popup_manager, popup_done());
+
+    // Create a new toplevel, which will dismiss the popups
```

Comment:
As far as I can tell this is *not* guaranteed by the protocol.

Indeed, I don't think that the behaviour this test expects is guaranteed by the protocol:

> When compositors choose to dismiss a popup, they *may* dismiss every
> nested grabbing popup as well. When a compositor dismisses popups, it
> will follow the same dismissing order as required from the client.
(Emphasis mine).

I don't think that the compositor is required to dismiss any but the topmost popup, *but* if it dismisses any others it must do so in the specified order. (This seems like maybe a spec bug, and I'll file a couple once I recover(?!) my Freedesktop login).

I *think* it's guaranteed that clicking somewhere outside the surface will dismiss the popup, so maybe the test should be doing that?

Reply:
I think it poorly worded, but I'm not sure your interpretation is a reasonable behaviour.

I think it a spec bug and that the intent is "they _must_ dismiss every nested grabbing popup as well"

Did you log a bug?